webodf/lib/gui/TrivialUndoManager.js

```diff
@@ -37,32 +37,124 @@ gui.TrivialUndoManager = function TrivialUndoManager(defaultRules) {
         domUtils = core.DomUtils,
         /**@type{?Element}*/
         initialDoc,
-        /**@type{!Array.<!ops.Operation>}*/
-        initialState = [],
+        /**@type{!StateTransition}*/
+        initialStateTransition,
         playFunc,
         /**@type{!ops.Document}*/
         document,
-        /**@type{!Array.<!ops.Operation>}*/
-        currentUndoState = [],
-        /**@type{!Array.<!Array.<!ops.Operation>>}*/
-        undoStates = [],
-        /**@type{!Array.<!Array.<!ops.Operation>>}*/
-        redoStates = [],
+        /**@type {!StateId}*/
+        unmodifiedStateId,
+        /**@type{!StateTransition}*/
+        currentUndoStateTransition,
+        /**@type{!Array.<!StateTransition>}*/
+        undoStateTransitions = [],
+        /**@type{!Array.<!StateTransition>}*/
+        redoStateTransitions = [],
         eventNotifier = new core.EventNotifier([
             gui.UndoManager.signalUndoStackChanged,
             gui.UndoManager.signalUndoStateCreated,
             gui.UndoManager.signalUndoStateModified,
+            gui.UndoManager.signalDocumentModifiedChanged,
             gui.TrivialUndoManager.signalDocumentRootReplaced
         ]),
         undoRules = defaultRules || new gui.UndoStateRules(),
-        isExecutingOps = false;
+        isExecutingOps = false,
+        /**
+        * Base for generating unique state ids
+        * @type {!number}
+        */
+        stateIdBase = 0;
+
+    /**
+     * @constructor
+     * @param {!number=} mainId
+     * @param {!number=} subId
+     */
+    function StateId(mainId, subId) {
```

Comment:
I would like to, but actually there is a dependency: `StateTransition` uses the `undoRules` set for the `TrivialUndoManager` object.  This could be solved by having each `StateTransition` also taking a constructor parameter `undoRules` but that seemed like unneeded data duplication. No measurements done, so perhaps not a real issue.

What do you think? Adding a new property to `StateTransition` worth the independency?